### PR TITLE
FcsDb macro update and SpinPool StTriggerFiltermaker request which seems to be lost

### PR DESCRIPTION
**Pull request by akioogawa**
FcsDb macro update and SpinPool StTriggerFiltermaker request which seems to be lost

### CODE REVIEW

**Comment by akioogawa** [comment]
I have no idea what CI Build / ROOT5 (pull_request)  is and what to do with it :P

**Comment by plexoos** [comment]
The cached layers in the docker image have been "lost" due to an update of the SL7 base image. It does not make sense to run CI now as it will hit the time limit anyway when trying to rebuild everything. I am rebuilding the image locally and will upload when ready.

File: StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx [hunk 1]
```diff
@@ -97,81 +95,98 @@ int StFcsTriggerSimMaker::Init(){
...
+	mTrgSim->stage_version[0]=2;
+	mTrgSim->stage_version[1]=1;
+	mTrgSim->stage_version[2]=6;
+	mTrgSim->stage_version[3]=3;
+    }
+
+    //Thresholds
+    mTrgSim->EM_HERATIO_THR = 32;
+    mTrgSim->HAD_HERATIO_THR = 32;
+    if(0){
```
Comment (klendathu2k) [inline StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Huh?  Why are so many thresholds behind an if (0)?  I assume these values are for previous years runs, and not current?  But if so, that means this code can only support one year's run at a time.  I realize this is in the spin pool (and thus not allowed in data productions)...

... but it does show up in embedding productions from time-to-time.  We need to start thinking about a solution to this issue, because at some point experts will not be available to remember which parts of the code need to be recompiled for what years.

Comment (klendathu2k) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Oh wait... didn't see this is in FcsTrigSimMaker...  So are these for different run periods?

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Ugh. This code is not supposed to be a part of this pull request. I keep forgetting that codes pushed to my GitHub will be added to pre-existing PR :P 
 Those thresholds need to be coming from RunControl->OnlineDB->OffllineDB. But Tonko & Jeff has NOT set up the first part yet and changing thresholds is editing a text file on daq computer right now. So no such DB entries exist on online DB yet, thus no copy in offline DB neither. So while we commissioning the FCS trigger during run21 and changing thresholds, I kept record of these thresholds in the code. So those are even different runs within an hour. We'll clean this up when actual DB is established. Until then, I won't touch those. I don't even need to push this to main STAR GitHub. But other people in FCS group need this code now for MC studies (and does not care about thresholds). So I'm asking them to get codes from my GitHub (git@github.com:akioogawa/star-sw). So I needed to push them. Do whatever you want on this file for STAR GitHub repo and I don't care.

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: There should be "remove this commit from the pull" button at each commit listed in this pull request page :P

Comment (veprbl) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: > There should be "remove this commit from the pull" button at each commit listed in this pull request page :P

Pull request, by definition, is a request to pull and merge a whole branch from your repository. You have full control over which commits are on the branch and which are not. In general it's not a good idea to make pull requests from your common work branch such as main (master) to not cause a situation like this one.

I would suggest you make a new PR with just the indended change on a separate branch. The instruction in https://github.com/star-bnl/star-sw/issues/35#issuecomment-867855609 should be applicable here as well.

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: git remote add upstream git@github.com:star-bnl/star-sw.git
git fetch upstream
git checkout upstream/main
git branch pr/fcsDetectorPosition_db.C
git reset --hard upstream/main
git cherry-pick ddb17db17b8f1f28068246805ff1750f8773c1be
em StRoot/StFcsDbMaker/macro/fcsDetectorPosition_db.C
git add StRoot/StFcsDbMaker/macro/fcsDetectorPosition_db.C
git commit
git push -u origin pr/fcsDetectorPosition_db.C

and ended up with a branch with all the commits from today again. I have no idea.

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: I believe I was closely following "Common workflow" from https://github.com/star-bnl/star-git-tools/blob/master/doc/README.md#common-workflow. (it misses add and commit too). If this is not "In general it's not a good idea", please update the document. This was a single command in CVS. This should not be so complicated using 10 lines (and using unreadable commit #s).

Comment (veprbl) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: > ```
> git remote add upstream git@github.com:star-bnl/star-sw.git
> git fetch upstream
> git checkout upstream/main
> git branch pr/fcsDetectorPosition_db.C
> ```

`git branch` creates a branch that points to the current commit, but it does not switch to said branch. You should have either followed it by `git checkout pr/fcsDetectorPosition_db.C`, or just used `git checkout -b pr/fcsDetectorPosition_db.C` that would create the branch and switch to it.

> ```
> git reset --hard upstream/main
> ```

This is not needed, as you already created branch while your `HEAD` was on the commit pointed by `upstream/main`.

> ```
> git cherry-pick ddb17db17b8f1f28068246805ff1750f8773c1be
> em StRoot/StFcsDbMaker/macro/fcsDetectorPosition_db.C
> git add StRoot/StFcsDbMaker/macro/fcsDetectorPosition_db.C
> ```

Well done so far.

> ```
> git commit
> ```

Last command should have been `git cherry-pick --continue`, but `git commit` kinda of works here. The only problem is that this commit didn't go to your "pr/fcsDetectorPosition_db.C" branch.

> ```
> git push -u origin pr/fcsDetectorPosition_db.C
> ```

At this point you've pushed your branch, but it should be still pointing to the same commit as upstream/main.

> 
> and ended up with a branch with all the commits from today again. I have no idea.

I see there are bunch of commits on https://github.com/akioogawa/star-sw/commits/pr/fcsDetectorPosition_db.C
but I don't understand how you could get to that state following the commands above.

Comment (veprbl) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: > I believe I was closely following "Common workflow" from https://github.com/star-bnl/star-git-tools/blob/master/doc/README.md#common-workflow. (it misses add and commit too). If this is not "In general it's not a good idea", please update the document.

I wasn't familiar with the document. The "Make changes locally." step seems to be a bit vague, an, indeed, it doesn't mention creating separate branches for different pull requests.

> This was a single command in CVS.

You are comparing to committing changes directly to trunk?

> This should not be so complicated using 10 lines (and using unreadable commit #s).

The basic use of git is very low level, but once you get some understanding of the concepts it is quite efficient to use.

The hashes are cryptic, but it's not like you need to remember them long term. Usually you just peek at `git log` and copy them before doing the operation and that's it. For everything else it's better use named branches or tags.

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: I have full command history and a log file. I had git status/log/branch --all etc in between. So it's quite long  (~1000 lines) so I edited out those lines which seems to me irrelevant. Sorry if I missed something important. 

If there is a better README, let me know. 

If this README remains to be the document I should give out to new students, please update "common workflow" with add & commit, how find commit# from long "git log" message, how to make a branch & cherry-pick and start a pull request on web page (can I do it from command line?) and what to do after (do I need git cherry-pick abort"? "git switch main" to go back?). 

While you are at it, "cvs commit" row in "Git and CVS" table at end should have those ~10 lines of git commands listed as well. And add "cvs checkout" row and corresponding lines (git clone & cd star-sw & git config core.sparseCheckout true & edit .git/info/ & git checkout, I guess 5 lines).

 I have no idea what trunk is. But I understood now that I need a quite bit of work even when I need to fix a typo in a comment line in a file of which I'm the owner.

Here is another try. Please let me know if anything is wrong here (I have no idea why this is bold)
----
 git remote add upstream git@github.com:star-bnl/star-sw.git 
 git fetch upstream
 git checkout upstream/main
 git branch [branchName]
 git checkout [branchName] 
 git log	 (search and find out commit#s you want for PR)
 git cherry-pick [commit#]
 git cherry-pick --continue
 git push -u origin [branchName]
 git cherry-pick --abort
 git switch main 
 (Go to https://github.com/[GitHub User Name]/star-sw)
 (Click "main" and type in [branchName] at Find and select it)
 (Click "Contribute"->"Open pull request")
----

Oh BTW, this is not a god place to ask. But while I have your attention, I have a instruction for new students on how to start working on FCS software at
  https://www.star.bnl.gov/protected/spin/akio/fcs/howto_MC_github.html
Please let me know if I have something wrong in section 2 & 3. Thanks!

Comment (plexoos) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: I am lost in all of these endless complaints mixed with irrelevant information and commands that don't make sense... Let's reset. What exactly are you trying to accomplish?

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Another question. If there are 3 commits in time A (file X) -> B (file X) -> C (file Y), and I want a PR including A & B but not C, do I cherry-pick A and B, or just B? 

What if timeline is A(x)->C(y)->B(x)? Does cherry-pick B includes A and C? Or does cherry-pick A & B exclude C?

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: plexoos, give me commands that make sense then.

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: And update the README

Comment (plexoos) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: > plexoos, give me commands that make sense then.

I'll give you commands if you tell me what you are trying to do

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: I just want to equivalent of "cvs ci fileA" while I have, or more importantly will have in future while the PR is still up, other changes/commits/pushes in my local and remote(origin? upstream? master? whatever you call git@github.com:akioogawa/star-sw)  repo, but not related to the fileA. 

Following veprbl's suggestion, I was trying to do that by using branch & cherry pick. 

Veprbl said "In general it's not a good idea to make pull requests from your common work branch such as main (master) to not cause a situation like this one". If what veprbl said is true, the "normal workflow" and "CVS vs Git" table in https://github.com/star-bnl/star-git-tools/blob/master/doc/README.md should be updated accordingly, or give us better document.

Comment (plexoos) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Here is what I recommend to do:

```
git clone git@github.com:akioogawa/star-sw.git
cd star-sw/
git remote add upstream git@github.com:star-bnl/star-sw.git
git fetch upstream main
git checkout -b newbranch upstream/main
<Modify fileA>
git commit <fileA>  -m "Log message"
git push -u origin newbranch
```

Then go to https://github.com/akioogawa/star-sw and create a new pull request from newbranch

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: This does not make sense to me. FileA is already modified, added, committed, pushed and is in the main branch of git@github.com:akioogawa/star-sw.git. 

Let me try again. Say my git@github.com:akioogawa/star-sw.git main branch looks like 

  Commit A (fileX)
  Commit B (fileY)
  Commit C (fileX)
  At this point, I want to start a PR for fileX with commit A & C, and file Y/Commit B should not be a part of the PR
  Future commit D (file Z)
  PR is still open with just A & B, unaffected by commit D

Since "You have full control over which commits are on the branch and which are not" (veprbl), I should be able to do this, right?

Comment (plexoos) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Yes, it is possible to cherry-pick specific commits from another branch onto the newbranch. In the commands I sent you replace

```
<Modify fileA>
git commit <fileA>  -m "Log message"
```
with
```
git cherry-pick <SHA of commit A>
git cherry-pick <SHA of commit C>
```

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: 3rd (4th?) try. Tell me this make sense:
 
          git remote add upstream git@github.com:star-bnl/star-sw.git
           git fetch upstream main
           git checkout -b [branchName] upstream/main
           git log (search and find out commit#s you want for PR)
           git cherry-pick [commit#1]
           git cherry-pick [commit#2]
           git cherry-pick [commit#3]
           git push -u origin [branchName]
           git cherry-pick --abort
           git switch main
           (Go to https://github.com/[GitHub User Name]/star-sw)
           (Click "main" and type in [branchName] at Find and select it)
           (Click "Contribute"->"Open pull request")

Comment (plexoos) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: Except the optional commands such as `git log` and `git switch`, and `git cherry-pick --abort` the purpose of which I don't understand, this looks identical to what I sent

Comment (plexoos) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: If you are lucky and there are no conflicts during cherry-picking then all of this might work. Otherwise, I'd suggest sticking to my original recipe with `<Modify fileA>; git commit <fileA>  -m "Log message"` which is as close as you can get to doing it CVS-way.

Comment (akioogawa) [reply StRoot/StSpinPool/StFcsTriggerSimMaker/StFcsTriggerSimMaker.cxx:118]: How do I get commit# w/o git log (other than going to web page)?
Don't I need to stop cherry-picking somehow? I got such message from git recommending abort.
Don't I need to switch branch after this, so that I don't push another code to PR unintentionally?
How do I work Git-way?
